tensorflow-frameworks/tensorflow-data/README.md

```diff
@@ -0,0 +1,220 @@
+
+Using the same example dataset from above, dataset elements can be extracted and
+used as follows:
+```java
+// Use default EagerSession
+Ops tf = Ops.create()
+
+// Dataset of (features, labels) from above
+Dataset dataset = Dataset.fromTensorSlices(tf, ... );
+
+// batch dataset elements into batches of size 2
+dataset = dataset.batch(2);
+
+Optmizer optimizer = ... // TF Optimizer
+
+for (List<Output<?>> batch : dataset) {
+    Operand<?> featureBatch = element.get(0);
+    Operand<?> labelBatch = element.get(1);
+
+    // Perform batch-wise computations on featureBatch and labelBatch
+    // e.g. computing model losses, running optimizers.
+
+    Operand<TFloat32> loss = myModelLoss(featureBatch, labelBatch);
+
+    optimizer.minimize(loss);
+    
+    ...
+}   
+
+```
+
+#### Graph Mode: OneShotIterator
+
+The above code will not work in graph mode, which requires the use of `Session`s
+to run the computations. In graph mode, datasets can be iterated over using the `DatasetIterator` abstraction, and a while loop.
+
+Once the iterator is initialized, repeated calls to `Session.run` will populate the components with new values, until all elements have
+been retrieved. After this, `Session.run` will result in an `IndexOutOfBounds` exception.
+
+Note that the make-iterator operation can be re-run to re-initialize
+the iterator, to iterate over the dataset a second time.
+
+```java
+try (Graph graph = new Graph()) {
+    // Graph mode Ops accessor
+    Ops tf = Ops.create(graph)
+
+    // Dataset of (features, labels) from above
+    Dataset dataset = Dataset.fromTensorSlices(tf, ... );
+
+    // batch dataset elements into batches of size 2
+    dataset = dataset.batch(2); 
+
+    // makeOneShotIterator() automatically adds the 
+    // iterator initializer (MakeIterator) Op to the Graph
+    // initializers list. Make sure to run `session.run(tf.init())`
+    // first!
+    DatasetIterator iterator = dataset.makeOneShotIterator();
+    List<Output<?>> batch = iterator.getNext();
+
+    Operand<?> features = batch.get(0);
+    Operand<?> labels = batch.get(1);
+
+    // Run additional computations on `features` and `labels`,
+    // e.g. computing model losses, instantiating Optimizers
+
+    Optimizer optimizer = ... // TF Optimizer 
+    Operand<TFloat32> loss = myModelLoss(features, labels);
+    
+    Op trainOp = optimizer.minimize(loss)
+
+    // instantiate graph-mode session
+    try (Session session = new Session(graph)) {
+        // Run graph initializers (and the iterator initializer)
+        session.run(tf.init());
+
+        // Iterate over dataset elements
+        while (true) {
+            try {
+                // Run training ops / fetch loss
+                List<Tensor<?>> outputs = session.runner()
+                    .addTarget(trainOp)
+                    .fetch(loss)
+                    .run();
+
+                ...
+
+            } catch (IndexOutOfBoundsException e) {
```

Comment:
I think we can get avoid having this exception as part of the normal workflow if we use `tf.data.iteratorGetNextAsOptional` instead of `tf.data.iteratorGetNext`, but that might require other changes as well in the graph so it is handle properly. 

Maybe the framework can automatically add what is required to do it? I did not went very far in my analysis, I just dropped the idea here so you can check if it make sense or not.

Reply:
Another solution could be to add a method called `runAllBatches()` in the `Session` or the `SessionRunner`, which accept a lambda from the user so he can intercept the output tensors after each run if desired. The `runAllBatches()` method would hide these intrinsic details of the loop that are not common for a framework library like TF Java (e.g. the `while (true)` and the `try/catch` statements)

Reply:
What would be interesting about that idea of accepting a lambda to consume the output tensors is that the framework could take care of freeing them after the lambda call is completed (or the lambda could return `true` or `false`, depending if the tensors reference should remain valid or not). 

This way, we would avoid this annoying `try-with-resources` block when fetching the output tensors. Wdyt? and @Craigacp ?

Reply:
Thanks for the suggestion Karl! I experimented with the `iteratorGetNextAsOptional`; it does make eager-mode iteration cleaner!

Using a `DatasetIterator` directly for eager-mode iteration, the syntax looks like
```
Dataset dataset = Dataset.fromTensorSlices(tf, tensors, dataTypes);
DatasetIterator iterator = dataset.makeOneShotIterator();
        
DatasetOptional nextOptional = iterator.getNextAsOptional();
while (nextOptional.hasValue().data().getBoolean()) {
    List<Output<?>> outputs = nextOptional.getValue();
    // use outputs in some computation.
    nextOptional = iterator.getNextAsOptional();
}
```
though this is still wrapped by the `Dataset` Iterable, so that the for-loop can be used: 

`for (List<Output<?>> outputs : dataset) { ... }`

In graph mode, this will get tougher. In a graph session, calling `tf.data.optionalGetValue` on an optional that doesn't have a value throws an `IllegalArgumentException`. In the Python tests for `iteratorGetNextAsOptional`, you can see that this exception is expected there also (see the test [here](https://github.com/tensorflow/tensorflow/blob/d2ad4a5c4019b4f9f4b882e4933af964f912cea8/tensorflow/python/data/kernel_tests/optional_test.py#L374)). Looks like they are stuck with the same issue :). 

In graph mode, there isn't an easy way of making sure that we only make the `iteratorGetNextAsOptional` call if `tf.data.optionalHasValue` returns true.

Fetching `hasValue` in it's own session.run doesn't work either, as every session.run will advance the iterator position.

The only other thing I can think of is to use control flow ops. The two options there would be to

1. Use a `switch` to introduce a branch on `hasValue`; I looked into this but I think there are some constraints that make this harder than it looks.

2. Use a graph `whileLoop` to run the entire iteration as a single sesion.run()... I think that may be a nice way of implementing the `runAllBatches` you mentioned, Karl; but still within a single `session.run()`, but until we have a working graph while loop, this will be very hard.

I'd love to look into both of these; but I think they are complexities we may not want to put on the user yet. I'm inclined to stick with the `IndexOutOfBoundsException` in graph mode for now. Even if it's sub optimal Java iteration style, the pattern is familiar to tensorflow data users.

Reply:
Ok, the idea behind the `runAllBatches` method would be to continue using `IndexOutOfBoundException` to stop the epoch iteration but to do it internally, something like this:

In `Session.Runner`:
```java
// consumer receives in parameter the batch number and the result of the run for this batch,
// and returns a boolean telling if if it is safe to release the fetched tensors (true) or not
void runAllBatches(BiFunction<Integer, Session.Run, Boolean> consumer) {
  int batch = 0;
  while (true) {
    try {
      Run run = run();
      if (consumer.apply(batch++, run)) {
        run.outputs.forEach(Tensor::close);  
      }
    } catch (IndexOutOfBoundException e) {
      break; // end of the loop   
    }
  }  
}
```
In user space:
```java
void trainGraph(Graph g) {
  try (Session s = new Session(g)) {
    s.runner().addTarget(trainOp).fetch(loss).runAllBatches((batch, run) -> {
      // print loss, etc.
      return true;
    });
  }
}
```
And don't have to restrict this signature for dataset batches only, we could add a `consumer` parameter to other `run` method as well. What I like is that we remove from the user the burden of releasing the fetched tensors himself, plus the new one of catching an exception to stop the training loop. I just feel that we are expecting too much knowledge from the user to be able to use our framework properly, while Java developers are more used with "don't-worry-we'll-take-care-of-it" libraries :). 

In comparison, the user code without the proposed changes would look like this.
```java
void trainGraph(Graph g) {
  try (Session s = new Session(g)) {
    int batch = 0;
    while (true) {
      List<Tensor> tensors;
      try {
        tensors = s.runner().addTarget(trainOp).fetch(loss).run().outputs;
        // print loss, etc.
        ++batch;
      } catch (IndexOutOfBoundException e) {
        break;
      } finally {
        tensors.forEach(Tensor::close());
      }
  }
}

Reply:
Hey @karllessard , this could be quite interesting! I think the mechanics will get quite difficult, though. The primary operation users will want to do with an iteration through all batches is to aggregate a value emitted from the batch (e.g. batchLoss, batchAccuracy) over the entire epoch.

With a lambda on each batch, it becomes more complicated to perform a "epoch loss update" like this, since within a lambda you can't modify non-final variables. So something like this won't be possible without additional machinery:

```
try (Session session = new Session(g)) {
    for (int epoch = 0; epoch < 10; epoch++) {
        int epochLoss = 0;
        session.runner()
            .addTarget(trainOp)
            .fetch(loss)
            .runAllBatches((batch, run) -> {
               epochLoss += run.loss; // not easily possible
               return true;
        });
    }
}
```
In the interest of simplifying things where possible, I had an idea for introducing a `Session.Result` (like the `Session.Run` you mention) to wrap the `List<Tensor<?>>` output from a `run()` call. See a potential implementation [here](https://github.com/tensorflow/java/compare/master...dhruvrajan:session-result?expand=1). 

We can reduce the boilerplate involved in managing / unwrapping this list, so that in a standard training loop, we get:

```
try (Session session = new Session(g)) {
    for (int epoch = 0; epoch < 10; epoch++) {
        int epochLoss = 0;
        int epochAccuracy = 0;
        try (Session.Result result = session.runner()
                    .addTarget(trainOp)
                    .fetch(loss)
                    .fetch(accuracy)
                    .fetch(prediction)
                    .run()) {
            
            // Outputs are retrieved from `results` with a sequence 
            // of pop() calls matching the order of fetch() calls.

            // Simple helper functions to fetch scalar primitive values.
            float batchLoss = result.popFloat();
            float batchAccuracy = result.popFloat();

            // Just an example, in case we want to retrieve a large tensor
            Tensor<TFloat32> batchPrediction = result.pop(TFloat32.DTYPE);

            epochLoss += batchLoss;
            epochAccuracy += batchAccuracy;
    }
}
```
This adds savings in a few places:
- Simplifies the arguments to the try-with-resources block; only one statement is needed, since the entire `Session.Result` object is `Autocloseable`
- Simplifies the call chain required to extract typed tensors (and even more for scalar values)
- Eliminates the need to access individual indices of an output list; calling `pop` in the same order as items were fetched is easier than keeping track of indices.

With the current setup we'd need to have

```
try (Session session = new Session(g)) {
    for (int epoch = 0; epoch < 10; epoch++) {
        int epochLoss = 0;
        int epochAccuracy = 0;

       List<Tensor<?>> outputs = session.runner()
                    .addTarget(trainOp)
                    .fetch(loss)
                    .fetch(accuracy)
                    .fetch(prediction)
                    .run();

        try (Tensor<TFloat32> lossTensor = outputs.get(0).expect(TFloat32.DTYPE);
              Tensor<TFloat32> accuracyTensor = outputs.get(1).expect(TFloat32.DTYPE);
               Tensor<TFloat32> predictionTensor = outputs.get(2).expect(TFloat32.DTYPE);) {

            float batchLoss = lossTensor.data().getFloat();
            float batchAccuracy = accuracyTensor.data().getFloat();

            epochLoss += batchLoss;
            epochAccuracy += batchAccuracy;
    }
}
```
What do you think?

Reply:
I was afraid that trying to modify final variables in the lambda could be a problem, your example is a good one... like you said, there are ways to go around this (e.g. using `AtomicInteger`) but maybe not as trivial as we would like.

Overall, I like your proposition (I think we should keep `Session.Run` and just make it auto-closeable and add the `pop*` methods, as this class already exists). For sure, it looks better than it is right now. The "expected exception" mechanism is still required though... I'd be curious to know what other members think about this.

Reply:
Ahh, yes I didn't realize `Session.Run` already exists; yup it makes sense to use it!

> The "expected exception" mechanism is still required though
Yup it will be for now ... Definitely would love to keep exploring alternative mechanisms here! If everyone uses Keras model fitting, no one will be the wiser! ;)

Reply:
I've done something similar to `Session.Run` in the onnx-runtime Java API (https://github.com/microsoft/onnxruntime/blob/54bbbb78ae3eefe69bd27f71dcc5affa977842c2/java/src/main/java/ai/onnxruntime/OrtSession.java#L613) which works fairly well, though it still feels a bit suboptimal (which I think is mainly because I'm fighting the type system a bit).

Given the purpose of the run all batches code would be to iterate a dataset then we should try to make it safer. What happens if a `DatasetIterator` isn't defined on the graph? Also I think that if we go down this path then we could modify the function to return either `T` and let the user unpack it how they wish, or some sort of `ComputationResult<T>` which is a tuple of a `T` and a boolean if we want to persist the tensors outside the run. I think that not closing the tensors in this kind of use case for training wouldn't be that likely, so I'm not sure we'd need to support it.

Does the `IndexOutOfBoundsException` come from here (https://github.com/tensorflow/java/blob/9fe9d87a454308e020b57839abf2e061f04c5a1e/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/c_api/AbstractTF_Status.java#L80)? I've not had chance to rebuild TF with your branch yet, so I can't quickly tell.

Reply:
It must come from there yes, that’s where TF statuses returned by the C API are mapped to exceptions.

It now sounds to me @dhruvrajan that we should go with your original proposal for now and work to improve the `Session` API as a separate topic that requires more thinking on how to manage native resources in general. It will probably end up wrapping existing mechanisms so it shouldn’t break too much the backward compatibility with what we are doing here. What do you guys think?

Reply:
That sounds good to me, it's a great idea to have a specific investigation into improving the Session / other native resources APIs. A helpful thing to focus on there might be some form of (potentially unified) management of "results lists": `List<Tensor<?>>` (graph mode) and `List<Output<?>>` (eager mode). The `pop()` methods in `Session.Run` I proposed above help some for Graph Sessions; it'd be awesome to find a way to do the same (maybe duplicate that?) for eager mode.

We can move forward with the current proposal, and hopefully update it once we think of something more user-friendly!

Reply:
I would prefer that this API be disabled for Graphs rather than have anything that caught a runtime exception in it. The `IndexOutOfBoundsException` can be thrown in multiple places from the `session.run()` call, as it returns a `List<Tensor<?>>` which throws `IndexOutOfBoundsException` if the `list.get()` call is out of bounds. This means while a user is developing their code it will just silently do nothing if they forget to add a fetch target, as the exception thrown from `list.get(0)` will cause the loop to break. Similarly if the graph ever returns a different number of tensors than requested, it will cause the loop to silently break out.

Reply:
Hmm... A possible solution (though it's high touch) could be to create a class of "TensorFlow Exceptions" in Java, and map to those instead of standard Java Exceptions in `Abstract_TF_Status`.

This way, we'd know that by catching a standard Java exception, we're handling a problem resulting from Java code, and by catching a TensorFlowException, we're handling an issue potentially caught at the C level (like a tf data out of range error)

Even without this though, I don't think the ambiguity `IndexOutOfBounds` introduces is a huge problem, since it's nearly the same ambiguity Python users have, unless I'm misunderstanding...

What do you think @Craigacp ?

Reply:
Actually, in Python, the error that is caught is a `tf.errors.OutOfRangeError`

Reply:
I stand by my previous statement on runtime exceptions, and it's really that exceptions shouldn't be control flow (e.g. Effective Java 3rd Ed Item 69). I'm a bit less worried by it being a checked exception out of the run method that we force users to deal with, but that's horrible for everyone who doesn't use the dataset api.

Reply:
I agree with Adam that the Java solution shouldn’t be based on the Python one since they are two very different languages with different type of users. And in Java, an exception is an unexpected error. 

So it might make sense for user jumping from TF Python to Java to see this mechanism in place but for someone who just starts from Java, it looks like a flaw in the library design. 

What about `Runner.runNextBatch` then which returns true if the next batch did not throw (internally) an exception and false if it did?Then you can retrieve the last `Run` metadata by getting `Runner.lastRun()` with a `try-with-resources` block. i.e.

```java
Runner runner = s.runner().fetch(loss);
while (runner.runNextBatch()) {
  try (Run r = runner.lastRun()) {
    // play with data
  }
}
```
or something like that (I’m on my phone so it’s not the best tool for writing a complete solution...)

Reply:
That's fine by me. I think we'd need to hammer out some scoping issues (i.e. when the tensors get closed that the runner holds), and also what the semantics of that method is when there is no datasetiterator defined. I was also considering making the datasetiterator on the Java side add a boolean flag saying when it wasn't exhausted, and we could check for that.

More generally what are the semantics of the iterator if you save the graph to disk and load it back up again? And do we want it to work on iterators that were embedded by python graphs (as those wouldn't have the boolean flag).

Reply:
I agree @karllessard, catching the exception internally is probably the best way to limit the effect of this on users. 

The `runNextBatch()` / `lastRun` flow can work, but needs some slightly complex state management in `Runner`. Instead of keeping track of the state of previous results in the `Runner` class, we could have it return an `Optional<Run>`:

```
Runner runner = session.runner().fetch(loss);

Optional<Run> nextBatch;
while ((nextBatch = session.runner().runNextBatch()).isPresent()) {
   try (Run run = nextBatch.get()) {
       // play with data
    }
}
```

After experimenting more, though, I think that we can do much better using streams, and allowing the ability to "take" from, "limit", and "map" over a stream of session.run() calls.

We can introduce a method `repeat()` in `Session.Runner` that returns an "infinite" stream of `Session.Run`. To handle the iteration, the stream will end once we catch an `IndexOutOfBoundsException`.

```
/**
 * Creates a stream of `Session.Run` objects which can be
 * processed to retrieve the results of repeated calls
 * to Session.Runner.run().
 *
 * Each item in the stream contains the results corresponding
 * to "fetched" tensors from this Runner.
 *
 * When the graph contains a dataset iterator, the stream
 * will end once the iterator has reached the end of
 * its iteration.
 *
 * To process only a prefix of this stream, a limit can manually
 * specified using Stream.limit( ... ). Other Stream operators
 * can also be used.
 *
 * @return A Stream of Session.Run items containing results
 *         of repeated calls to Session.Runner.run()
 */
public Stream<Session.Run> repeat();
```

We can then use `map()` with a lambda to process batch updates, similarly to @karllessard's earlier proposal. While we still can't update primitive values, by creating a `Logs` object
containing the fields we want to extract, users can quite cleanly retrieve values that they
want to save.

```
class EpochLogs {
    float loss = 0;
    float accuracy = 0;
    int batches = 0;
}
``` 
With such a class, the training loop becomes quite simple:

```
// Run training loop
for (int i = 0; i < EPOCHS; i++) {
    // reset iterator object
    session.run(initTrainIterator);
    
    EpochLogs epoch = new EpochLogs();
    session.runner()
        .addTarget(optimizerTargets)
        .fetch(trainLoss)
        .fetch(trainAccuracy)
        .repeat() // Create a stream that ends once the iterator is exhausted.
        .forEach(outputs -> {
            epoch.loss += outputs.popFloat();  // here imagine output.get(...).expect(...).data().getFloat() . 
            epoch.accuracy += outputs.popFloat();
            epoch.batches++;
        });

    System.out.println("Epoch Accuracy " + i 
        + ": " + epochLogs.accuracy / epochLogs.batches);
}
```
There's a couple nice things about this:
- `Runner.repeat()` can be used for ANY tensorflow graph, even if it doesn't explicitly use
   `Dataset` ops. 
- The stream iteration takes care of closing the tensors returned each batch.
- Users can use the extensive set of Stream operators to control their loops.
- Logical distinction between `Session` and `Dataset` is maintained.

It's a little toubling that we have to catch an `IndexOutOfBoundsException` to handle exception from `tensorflow-core`, I think it'd be a great idea in a follow-up PR to add a set of exceptions extending `TensorFlowException` , and map to these in `TF_Abstract_Status.java`, so that we at least know when we are catching an issue from `tensorflow-core`.

What do you think? I've quite come around to liking lambda-style iteration here. See [MNISTBasicGraphClassifier.java](https://gist.github.com/dhruvrajan/8d5b9c5683f88c965ad756c1b0202055#file-mnistbasicgraphclassifier-java-L149) for a more complete example.

Reply:
>More generally what are the semantics of the iterator if you save the graph to disk and load it back up again?

@Craigacp that's a pretty interesting point, I haven't tried it (not familiar with the graph-to-disk saving and loading). But since the iterator state is saved in the graph, saving to disk and re-loading *should* save the state of the iterator.

Reply:
Hey @dhruvrajan , I like it! I’m not too sure if we should provide the `EpochLogs` class or just let the user create it based on his needs but that’s a detail.

For auto-closing all tensors after each batch, that can do to begin with but at some point, we might allow the user to also retain a reference to one or more of them if he wants to. Reference counting added lately to JavaCPP might be helpful here.

Reply:
@karllessard Cool! Yeah, that makes sense, that'll be quite cool if we can let the user maintain the tensor objects when useful.

As part of the Keras library, we'll actually have to add these `Logs` classes anyway, for now I think it's ok to leave that to the user, since that's the only way to allow full flexibility with the fields.

Reply:
To clarify the proposed semantics for `session.runner().repeat()` are that it repeats until it hits an `IndexOutOfBoundsException`, which it silently swallows and the stream terminates? 

If so in addition to the case where it works for the dataset iterator, there are two more cases, one where it goes forever executing identical operations (i.e. feeding the same tensors and receiving the same output), and another where TF_OUT_OF_RANGE is thrown by something else in the TF code which silently ends the execution (possibly in the middle of a dataset iteration due to a coding error). It looks like TF_OUT_OF_RANGE (or equivalently OUT_OF_RANGE) get thrown from a variety of different places in the TF runtime, not just from exhausted dataset iterators. This latter case is worrying to me as it could lead to situations which are very difficult to debug as we expect it to do one thing and silently eat the error message when it died for some other reason.

Also, the stream API removes control from the user of when the computation happens. This is usually not an issue as the laziness usually helps the compiler orchestrate things better, but TF computations are big and not as thread safe as we'd like them to be, because they modify mutable state. The spliterator built out of the iterator should have the `IMMUTABLE` and `ORDERED` characteristics set on it, which should prevent some of the oddities. Fortunately the `IteratorSpliterator.trySplit()` method will buffer elements of the underlying iterator into an `ArraySpliterator`, triggering TF computation sequentially, so if someone does call `stream.parallel()` on the output it should have the correct semantics. This computation happens up front when the stream is polled, rather than when the elements of the stream are actually processed by the end user, so that might be a little weird for them.

Reply:
Yup that's correct, the stream will terminate when the `IndexOutOfBoundsException` is caught.

> one where it goes forever executing identical operations (i.e. feeding the same tensors and receiving the same output)

To handle this case, users can specify a limit on the number of iterations with `Stream.limit`. In Java9+, they can also use `Stream.takeWhile`, which is much more flexible.

> and another where TF_OUT_OF_RANGE is thrown by something else in the TF code which silently ends the execution

Definitely, his is an unfortunate property of the `TF` runtime. I think the best we can do is make sure we treat these separately from other Java exceptions (i.e. not map this error to `IndexOutOfBounds`). This will a problem that exists across every TensorFlow system that uses the standard TF runtime.

I think that's why the "safest" thing to do is to surface the error to the user,  allow them to handle it, look at the actual error message (which states that the iterator is exhausted), etc, to distinguish it from other errors while debugging. Users can still use the `while` loop and catch syntax if they prefer.

But there doesn't seem to be a both safe and easy way of providing an API which hides the exception, so that's the tradeoff to make here. It sounds like we are clear that we should provide an API which doesn't require exception-handling on the user side; I think this would be a decent way of making that tradeoff, while providing additional benefits to the user.

Reply:
Ahh these are really interesting points, we will have to pay close attention to the stream semantics.

> the stream API removes control from the user of when the computation happens

It's true that the user won't be making the individual `session.run()` calls; but apart from this, but I don't think it would cause any major problems / slowdowns? Do you know of cases where this could cause an inadvertent performance issue?

> this computation happens up front when the stream is polled, rather than when the elements of the stream are actually processed by the end user

Hmm... I'm not too familiar with `Stream` internals, what do you mean by `polling` vs `processing` by the user? Is it that all batch session.run() calls will occur up front, before any of the user lambdas are run as a separate stage over the `Run` objects? I think that would be ok, as far as I know that's expected stream behavior and hopefully doesn't result in too much confusion.

Do you see any cases where it would be useful to run some tests, to make sure that this doesn't result in actual bugs?

Reply:
If they call `stream.parallel()` and then some terminal operation, all the computation will happen in the single thread dividing up the work for the parallel workers, as it happens on the call to next in the iterator and that's called repeatedly by `IteratorSpliterator.trySplit()` as it builds up a chunk to hand to an `ArraySpliterator`. I'm referring to this as polling the stream, where it chunks the work up, rather than processing (i.e. when the individual workers start executing the chain of operations leading to the terminal one). It's not quite the right terminology, but to talk about it precisely requires going through the internals in some detail.

It's not really expected behaviour that the computation happens while the stream is being chunked, as it tends to mess with the sizing internals of the Spliterator (e.g. it uses an increasingly large buffer before chunking off the work, and this can cause issues with IO operations. I had to implement my own spliterator to get around this in my Java word2vec implementation).

I feel this is trying to paper over an API that doesn't really work in Graph mode, especially in Java. Unless we know for sure that this code can iterate over a graph saved out from python using Python's dataset iterator, then I'm not sure what the benefit is of forcing Java to behave like this. Maybe we should build our own iteration operation that exposes the necessary checks that we can tightly monitor from the Java side? Either by wrapping the current dataset iterator ops with something that can grab the OUT_OF_RANGE status exactly where it occurs in only the iterator exhausted case, or by building something that slices the tensors by hand and serves up another one each time entirely from primitive ops we control (with appropriate teaching of the Session API so it has appropriate entry points for pumping an iterator, that are strongly typed and prevent users from making mistakes).

Reply:
>  wrapping the current dataset iterator ops with something that can grab the OUT_OF_RANGE status exactly where it occurs in only the iterator exhausted case

This would certainly be great; but it requires a huge amount of effort, not just on the Java side, but within the C++ runtime. That seems too complicated for the scope of this PR.

We have considered a few alternatives; I think it'd be best to choose one, move forward with it, and revisit the semantics of graph-mode dataset iterators when we have better iterator support in Java.

1. Standard tf.data syntax: Propagate the `IndexOutOfBounds` exception to the user so it can be caught in a `while(true)` loop. The user handles success checking (catching the error), as well as closing the returned result tensors.

2. Catch the `IndexOutOfBounds` error internally, and return an `Optional` (or boolean) indicating success at each iteration to the user. The user has to manage closing the result tensors.

3. Make repeated calls to `session.run()` internally, and allow the user to process results with a lambda, catching the error internally, and closing tensors automatically. This can be either a single function in `Session`, or accomplished via streams (I think, more flexible and usable).

Let's pick one of these and revisit this later, as appropriate? I prefer the first, as it's the simplest, despite the obvious shortcomings.

@karllessard if you think it would be helpful, I can consolidate each of these, and send out a poll in the mailing list.

Reply:
Re: `Stream.parallel`, those are great points, and if we go that route, we should restrict the stream to run sequentially (return null from `trySplit`, or override `parallel()`, perhaps).

Reply:
If we go with option 1, then at least we should make a new subclass of `RuntimeException` called `TFOutOfRangeException` and modify the TFStatus check to throw that rather than `IndexOutOfBoundsException` along with the necessary documentation changes. We should not throw `IndexOutOfBoundsException` in this case as it's semantically different to the `TF_OUT_OF_RANGE` "error".

Reply:
Also any of these approaches should be documented with big warning flags around them saying this is going to change when we figure out a better way of doing it. I really really don't want to be stuck supporting this API for Graphs for a long time.

Reply:
Option 1 is for sure the ugliest but the one having less impact on the `Session` API (in fact, it has none), therefore if we want to move that PR forward, we can go with this one and build something nicer on top of it at a later time.

In my [initial proposal](https://github.com/tensorflow/java/pull/30#discussion_r407256981) for handling the result in a lambda and auto-closing the tensors, there were no streams involve therefore no possible awkward behaviour with `Stream` like @Craigacp described (and which I did my best to understand :) ). Maybe we should go back to something like this then? 

But again, if we are not ready to do it, this could be done later if we go with Option 1 first, without breaking backward compatibility.

Reply:
...quite a thread guys, only to go back to the original solution 😄

Reply:
> Option 1 is for sure the ugliest but the one having less impact on the Session API (in fact, it has none)

Yup I totally agree. This seems to be safest, as it doesn't run the risk of hiding other errors from the user, unrelated to dataset iteration (like Adam mentioned, even if we introduce additional TensorFlow exceptions), which could make things pretty near impossible for the user to debug. That's my thinking here; better to not take this risk, until we are sure we have a better solution.

> Maybe we should go back to something like [lambdas] then?
Syntax-wise, I do actually really like the lambda-iteration approaches; both with and without streams, the resulting code is quite clean. Would love start introducing this separately! 

The thing that isn't yet (but can be) solved here is having an exit condition definable in the lambda. In `runAllBatches` we wait for the `IndexOutOfBoundsException; in Java 11 we have `takeWhile` which can help solve some of these issues. Starting from the `repeat` abstraction we could add methods like `repeatWhile(predicate)`, `repeatFor(iterations)`, or `repeatUntilFailure(consumer)` etc... Even if we implement it without streams, we can get some nice syntax.

Reply:
I think we should break compatibility in that we change what exception is thrown by `TF_OUT_OF_RANGE` to a new thing subclassing `RuntimeException`. Remember one of the issues with the blind catching of the `IndexOutOfBoundsException` is that if the user does `list.get(1)` on the `List<Tensor<?>>` returned by `session.runner().run()` it will silently swallow the `IndexOutOfBoundsException` thrown by `list.get()`.

Reply:
@Craigacp totally agree here, like I mentioned before, I think this will be a great safety improvement.

I can investigate this; I wonder how many places in our code are expecting `IndexOutOfBoundsException`; hopefully not too many.

Reply:
> I can investigate this; I wonder how many places in our code are expecting IndexOutOfBoundsException; hopefully not too many.

I guess none? All `IndexOutOfBoundException` that we might catch in the core API are the real runtime exceptions, I don't think there is any coming from `TF_Status` mapping... but I didn't look neither.

Reply:
Man, this truly is an epic thread. I'll save the link! :)